# Исследование надежности заемщиков


## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [2]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [3]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [5]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [6]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()
#  Альтернативный вариант  
# data['total_income'] = data['total_income'].fillna(data.groupby('income_type')['total_income'].transform('median'))

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [7]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [8]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [9]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [10]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [11]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [12]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [13]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [14]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [15]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [16]:
data.duplicated().sum()

71

In [17]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [18]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [19]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [20]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [21]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [22]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

В буквальном смысле "зависимость" определяют другими стат. методами. 

Тут мы можем увидеть максимум какие-то "ковариации-корреляции"

Определимся с двумя гипотезами и контр-гипотезами

1. Люди с детьми более ответственны и чаще возвращают кредиты (контр-гипотеза, соответственно наоборот)
2. Чем больше детей, тем хуже возвращают кредиты (и наоборот)





Посмотрим на дата-фрейм

Сейчас нас интересуют колонки children и debt

In [23]:
data.head()

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose total_income_category          purpose_category  
0               покупка жилья                     B  операции с недвижимостью  
1     приобретение автомобиля                     C    операции с автомобилем  
2               покупка жилья                     C  операции с недвижимостью  
3  дополнительное образование                     B     получение образования  
4             сыграть свадьбу                     C        проведение свадьбы

*Проверяем первую гипотезу*

In [24]:
# Применяем loc и округляем для удобства
deby_with_ch = round(data['debt'].loc[data['children'] == 0].mean(), 3)
deby_without_ch = round(data['debt'].loc[data['children'] != 0].mean(), 3)

print('Среднее кол-во просрочек у заемщиков без детей:', deby_with_ch)
print('Среднее кол-во просрочек у заемщиков с детьми:', deby_without_ch)

Среднее кол-во просрочек у заемщиков без детей: 0.075
Среднее кол-во просрочек у заемщиков с детьми: 0.092


In [25]:
# Как можно заметить, у заемщиков с детьми просрочек больше, но не сказать чтобы сильно больше
# 1-я Гипотеза не подтверждается (хотя корректней проводить стат. тест)

*Проверяем вторую гипотезу*

In [26]:
# Применяем группировку умножаем и округляем для удобства
round(data.groupby('children')[['debt']].mean(), 4)*100

debt
children      
0         7.54
1         9.23
2         9.45
3         8.18
4         9.76
5         0.00

Как можно заметить, чаще всего не возвращают кредиты люди с 4-мя детьми, но разница между группами не слишком большая. 
А у людей с 5-ю кредитами нет вообще просрочек? Сделаемсводнут таблицу и посмотрим кол-во 


In [27]:
# Посчитаем кол-во, а не среднее
# round(data.groupby('children')[['debt']].count(), 2)

In [28]:
# Посмотрим группировками
debt_from_children = pd.DataFrame() # создаем пустой датафрейм
debt_from_children['count_children'] = data.groupby('children')['debt'].count()
debt_from_children['sum_children'] = data.groupby('children')['debt'].sum()
debt_from_children['Просрочки %'] = round((debt_from_children['sum_children'] / 
                                               debt_from_children['count_children']*100), 2) 
debt_from_children

count_children  sum_children  Просрочки %
children                                           
0                  14091          1063         7.54
1                   4808           444         9.23
2                   2052           194         9.45
3                    330            27         8.18
4                     41             4         9.76
5                      9             0         0.00

Другие варианты группировок

In [29]:

# chil_debt=data.groupby('children').agg({'debt':['count','sum']})
# chil_quality=chil_debt['debt']['sum']/chil_debt['debt']['count']*100

Заемщиков с 5-ю детьми слишком мало. Да и с 4-я детьми тоже не много. Основные кредито-братели это бездетные

In [30]:
# Делаем сводную таблицу
data_ch_pivot = data.pivot_table(index='children', values='debt', aggfunc=['count', 'sum', 'mean'])


data_ch_pivot['Доля задолженности %'] =  data_ch_pivot['sum'] / data_ch_pivot['count']

# округлим и умножим на 100
data_ch_pivot['Доля задолженности %'] =  round((data_ch_pivot['Доля задолженности %']), 4)*100

# Переименуем столцбы для красоты 
data_ch_pivot.columns = [('Кол-во', 'count'), ('Сумма', 'sum'), ('Среднее', 'mean'), ('Доля задолженности %')]

print(data_ch_pivot)


          (Кол-во, count)  (Сумма, sum)  (Среднее, mean)  Доля задолженности %
children                                                                      
0                   14091          1063         0.075438                  7.54
1                    4808           444         0.092346                  9.23
2                    2052           194         0.094542                  9.45
3                     330            27         0.081818                  8.18
4                      41             4         0.097561                  9.76
5                       9             0         0.000000                  0.00


**Вывод:** 
В среднем, заемщики с детьми чаще допускают просрочки по кредитам, из заемщиков с детьми чаще всего просрочки встречаются у семей с 2-я и 4-мя детьми, однако группа заемщиков с 4-мя и 5-ю детьми небольшая, поэтому нерепрезентативна. Таким образом, самые ненадежные заемщики - это люди с двумя детьми

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

Для разнообразия, используем другой метод


In [31]:
 deb_fam_stat = data[['family_status', 'debt']]
 deb_fam_stat_grouped =  deb_fam_stat.groupby('family_status').count()
 deb_fam_stat_grouped['number'] =  deb_fam_stat.loc[deb_fam_stat['debt'] == 1].groupby('family_status').sum()
 deb_fam_stat_grouped['mean, %'] =  deb_fam_stat_grouped['number'] /  deb_fam_stat_grouped['debt'] * 100  # умножил на 100
 deb_fam_stat_grouped.sort_values('mean, %', ascending=False)

debt  number   mean, %
family_status                                 
Не женат / не замужем   2796     273  9.763948
гражданский брак        4134     385  9.313014
женат / замужем        12261     927  7.560558
в разводе               1189      84  7.064760
вдовец / вдова           951      63  6.624606

**Вывод:** 

Как можно заметить, люди не состоящие в браке или живущие совместно (то есть в "гражданском браке") чаще склонны к просрочке кредитов
Меньше всего просрочка кредитов у людей в разводе и вдоцов/вдов, однако кол-во таких заемщиков относительно небольшое

Проще говоря люди, имевшие опыт семейных отношений, более добросовестно относятся к выплате кредитов

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [32]:
# На этот раз попробуем посчитать циклом

In [33]:
for i in data['total_income_category'].unique():
    print('доля невозвращенных кредитов в срок в категории:', i, 'составляет', \
    round(100*data['debt'].loc[data['total_income_category'] == i].mean(), 2))
    



доля невозвращенных кредитов в срок в категории: B составляет 7.06
доля невозвращенных кредитов в срок в категории: C составляет 8.5
доля невозвращенных кредитов в срок в категории: D составляет 6.02
доля невозвращенных кредитов в срок в категории: E составляет 9.09
доля невозвращенных кредитов в срок в категории: A составляет 8.0


In [34]:
# Посмотрим сколько всего в каждой категории
data['total_income_category'].value_counts()


C    15921
B     5014
D      349
A       25
E       22
Name: total_income_category, dtype: int64

In [44]:
# сделаем простую сводную таблицу
data_ch_pivot = data.pivot_table(index='total_income_category', values='debt', aggfunc=['count', 'sum', 'mean'])


data_ch_pivot['Доля задолженности %'] =  data_ch_pivot['sum'] / data_ch_pivot['count']

# округлим и умножим на 100
data_ch_pivot['Доля задолженности %'] =  round((data_ch_pivot['Доля задолженности %']), 4)*100

# Переименуем столцбы для красоты 
data_ch_pivot.columns = [('Кол-во', 'count'), ('Сумма', 'sum'), ('Среднее', 'mean'), ('Доля задолженности %')]

display(data_ch_pivot)

(Кол-во, count)  (Сумма, sum)  (Среднее, mean)  \
total_income_category                                                   
A                                   25             2         0.080000   
B                                 5014           354         0.070602   
C                                15921          1353         0.084982   
D                                  349            21         0.060172   
E                                   22             2         0.090909   

                       Доля задолженности %  
total_income_category                        
A                                      8.00  
B                                      7.06  
C                                      8.50  
D                                      6.02  
E                                      9.09

**Вывод:** 
Первая тройка категорий по просроченности кредитов это
 - E (9.09%) с доходом от 30 до 50-и тысяч
 - C (8,50%) с доходом от 50-и до 200 тысяч
 - A (8.00%) от одного миллиона и выше
 
Можно сделать вывод, что первые 2 категории это физические лица, которые берут кредиты на текущие нужды. А категория А - предприниматели и руководители бизнеса

Заемщики с доходом от 200 тысяч до миллиона лучше справляются с долговой нагрузкой, а вот людям с меньшим доходом выплата кредитов в срок дается сложнее



#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [45]:
# Посчитаем сводной таблицей
target_credit = data.pivot_table(index = [ 'purpose_category'], values = 'debt', aggfunc = {'count','sum'})
target_credit = target_credit.rename(columns={'count':'total','sum':'target'})
target_credit['Соотношение'] = round(target_credit['target']/target_credit['total'], 4)
target_credit  = target_credit .sort_values(by='Соотношение', ascending=True)
display(target_credit)

total  target  Соотношение
purpose_category                                    
операции с недвижимостью  10751     780       0.0726
проведение свадьбы         2313     183       0.0791
получение образования      3988     369       0.0925
операции с автомобилем     4279     400       0.0935

In [46]:
# Альтернативный вариант 
# Посчитаем сумму и счет отдельно
count_credit = data.pivot_table(index=['purpose_category'], values='debt', aggfunc='count')
sum_credit = data.pivot_table(index=['purpose_category'], values='debt', aggfunc='sum')

# Объединяем в один дата-фрейм
target_credit = pd.merge(count_credit, sum_credit, on='purpose_category')

# Переименуем названия колонок
target_credit = target_credit.rename(columns={'debt_x': 'count', 'debt_y': 'sum'})

# Рассчитаем соотношение и сортируем
target_credit['Соотношение'] = round(target_credit['sum'] / target_credit['count'], 3)
target_credit = target_credit.sort_values(by='Соотношение', ascending=True)

# Печатаем
display(target_credit)

count  sum  Соотношение
purpose_category                                 
операции с недвижимостью  10751  780        0.073
проведение свадьбы         2313  183        0.079
операции с автомобилем     4279  400        0.093
получение образования      3988  369        0.093

**Вывод:** 
Как можно заметить, заемщики, берущие кредит на операции с недвижимостью и проведение свадьбы более ответственно относятся к выплате кредитов

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Невнимательность, Агрегация с разных баз данных и утеря в процессе, возможно для каких-то кредитов не нужна полная инофрмация о заемщике (например при зарплатных проектах банка)

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:*  медиана менее подвержена смещению и более показательна, когда данные распределены "Ненормально", не по нормальному распределению

### Шаг 4: общий вывод.

По итогам настоящего исследования, можно сделать следующий вывод: Наиболее надежный заемщик - это человек без детей, но с опытом жизни в официальном браке (а лучше даже разведенный или вдовец/вдова). С доходом выше 200 тысяч рублей, берущий кредит для целей операций с недвижимостью

С другой стороны, наиболее ненадежные заемщики - это люди без опыта жизни в браке, с детьми и берущие кредит на покупку автомобиля